cep-4.md

```diff
@@ -0,0 +1,134 @@
+
+The [initial proposal to add this plugin architecture](https://github.com/conda-incubator/ceps/blob/main/cep-2.md) has been [officially approved](https://github.com/conda-incubator/ceps/issues/23); this current CEP will discuss the specific way that the plugin mechanism will actually be implemented.
+
+## Specification
+
+Plugins in `conda` will integrate the "hook + entry point" structure by utilizing the [`pluggy`](https://pluggy.readthedocs.io/en/stable/index.html) Python framework. This implementation can be broken down via the following two steps:
+
+- Define the hook(s) to be registered
+- Register the plugin under the `conda` entrypoint namespace
+
+
+### Hook
+
+Below is an example of a very basic plugin "hook":
+
+_my_plugin.py_
+```python
+import conda.plugins
+
+
+@conda.plugins.register
+def conda_subcommands():
+    ...
+```
+
+### Entry point namespace utilizing a `setup.py` file
+
+Below is an example of an entry point namespace for the custom plugin function, decorated with the plugin hook shown above:
+
+_setup.py_
+```python
+from setuptools import setup
+
+setup(
+    name="my-conda-plugin",
+    install_requires="conda",
+    entry_points={"conda": ["my-conda-plugin = my_plugin"]},
+    py_modules=["my_plugin"],
+)
+```
+
+### Packaging via a `pyproject.toml` file
+
+
+Below is a different packaging example that configures `setuptools` using a `pyproject.toml` file:
+
+_setup.py_
+```python
+import setuptools
+
+if __name__ == "__main__":
+    setuptools.setup()
+```
+
+_pyproject.toml_
+```toml
+[build-system]
+requires = ["setuptools", "setuptools-scm"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "my-conda-plugin"
+description = "My conda plugin"
+requires-python = ">=3.7"
+dependencies = ["conda"]
+
+[project.scripts]
+my-conda-plugin = "my_plugin.module:function"
+```
+
+
+## Rationale
+
+This new `conda` plugin API ecosystem will bring about many benefits and possibilities, including but not limited to:
+
+- Custom subcommands
+- Support for packaging-related topics (_e.g._, virtual packages)
+- Development environment integrations (_e.g._, shells)
+- Alternative dependency solver backends
+- Experimental features that are not currently covered by `conda`
+
+As mentioned previously, an official plugin ecosystem will also enable contributors across the `conda` community to develop and share new features, thus bringing about more functionality and focus on the user experience.
+
+## Backwards Compatibility
+
+There are no expected backwards compatibility issues for this new feature.
+<!-- ??? -->
+
+## Alternatives
+
+The following lists alternative Python frameworks, libraries, and packages that were considered for use in this proposed `conda` plugin implementation:
+
+- [Hookman](https://github.com/ESSS/hookman)
```

Comment:
I would add pluggy here
```suggestion
- [Hookman](https://github.com/ESSS/hookman)
- [Pluggy](https://github.com/pytest-dev/pluggy)
```
thanks to @niccodemus for reminding me about ``pluggy`` :)